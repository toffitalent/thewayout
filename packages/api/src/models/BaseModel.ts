import { PaginateQueryBuilder, Serialize } from '@disruptive-labs/objection-plugins';
import { Model, ModelOptions, QueryContext } from 'objection';
import { knex } from '../db';

// Bind knex instance to objection
Model.knex(knex);

export class BaseModel extends Serialize(Model) {
  static QueryBuilder = PaginateQueryBuilder;
  QueryBuilderType!: PaginateQueryBuilder<this>;

  static useLimitInFirst = true;

  createdAt!: Date;
  updatedAt!: Date;

  async $beforeInsert(queryContext: QueryContext) {
    await super.$beforeInsert(queryContext);

    this.createdAt = new Date();
    this.updatedAt = this.createdAt;
  }

  async $beforeUpdate(opt: ModelOptions, queryContext: QueryContext) {
    await super.$beforeUpdate(opt, queryContext);

    this.updatedAt = new Date();
  }
}
