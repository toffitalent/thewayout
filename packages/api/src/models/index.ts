export * from './App';
export * from './AppClient';
export * from './AppToken';
export * from './AppUser';
export * from './BaseModel';
export * from './Client';
export * from './Employer';
export * from './Job';
export * from './JobApplication';
export * from './Rsp';
export * from './RspAccount';
export * from './RspInvitation';
export * from './serializers';
export * from './User';
export {
  CheckViolationError,
  ConstraintViolationError,
  DataError,
  DBError,
  ForeignKeyViolationError,
  NotFoundError,
  NotNullViolationError,
  raw,
  ref,
  UniqueViolationError,
  ValidationError,
} from 'objection';
