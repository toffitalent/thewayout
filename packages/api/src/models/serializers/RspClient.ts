import { Serializer } from '@disruptive-labs/objection-plugins';

export class RspClientSerializer extends Serializer {
  static attributes = ['id', 'rspId', 'status', 'caseManagerId', 'notes'];

  static relations = ['user', 'caseManager'];
}
