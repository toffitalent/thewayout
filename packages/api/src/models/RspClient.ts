import { Knex } from 'knex';
import { QueryContext, RelationMappingsThunk } from 'objection';
import * as Mail from '@two/mail';
import { offenseText, RspClientStatus, RspRole, supportText } from '@two/shared';
import { BaseModel } from './BaseModel';
import { Client } from './Client';
import { Rsp } from './Rsp';
import { RspAccount } from './RspAccount';
import { RspClientSerializer } from './serializers';
import { User } from './User';

export class RspClient extends BaseModel {
  static tableName = 'rsp_clients';
  static Serializer = RspClientSerializer;

  readonly id!: string;

  userId!: string;
  rspId!: string;
  caseManagerId?: string;
  status!: RspClientStatus;
  closedAt!: Date | null;
  notes?: string;

  caseManager?: RspAccount;
  rsp?: Rsp;
  user?: User;

  static relationMappings: RelationMappingsThunk = () => ({
    caseManager: {
      relation: BaseModel.HasOneRelation,
      modelClass: RspAccount,
      join: {
        from: 'rsp_clients.caseManagerId',
        to: 'rsp_accounts.id',
      },
    },

    rsp: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: Rsp,
      join: {
        from: 'rsp_clients.rspId',
        to: 'rsp.id',
      },
    },

    user: {
      relation: BaseModel.HasOneRelation,
      modelClass: User,
      join: {
        from: 'rsp_clients.userId',
        to: 'users.id',
      },
    },
  });

  async $afterInsert(queryContext: QueryContext) {
    await super.$afterInsert(queryContext);

    const newClient = await Client.query().where({ userId: this.userId }).first();
    const chosenRsp = await Rsp.query().findById(this.rspId);
    const rspOwner = await RspAccount.query()
      .select(['rsp_accounts.*', 'users.email', 'users.firstName', 'users.lastName'])
      .leftJoin('users', (join: Knex.JoinClause) => {
        join.on('rsp_accounts.userId', 'users.id');
      })
      .where({ rspId: chosenRsp?.id })
      .where({ role: RspRole.owner })
      .first();

    if (newClient && chosenRsp && rspOwner) {
      const clientUser = await User.query().findById(newClient.userId);

      const {
        email,
        firstName: ownerFirstName,
        lastName: ownerLastName,
      } = rspOwner as RspAccount & {
        email: string;
        firstName: string;
        lastName: string;
      };
      const {
        firstName,
        lastName,
        address,
        postalCode,
        city,
        state,
        phone,
        support,
        offense,
        justiceStatus,
        facility,
        expectedReleasedAt,
      } = newClient;

      await Promise.all([
        Mail.send(
          {
            to: {
              address: email,
              name: `${ownerFirstName} ${ownerLastName}`,
            },
            cc: { address: 'support@twout.org', name: 'Support' },
          },
          'RspApplicationSupport',
          {
            user: {
              name: `${firstName} ${lastName}`,
              email: clientUser!.email,
              address: `${address}, ${city}, ${state} ${postalCode}`,
              phone: phone || '',
              support: support.map((el) => supportText[el]).join(', '),
              offense: offense?.map((el) => offenseText[el]).join(', ') || 'N/A',
              justiceStatus,
              facility,
              expectedReleasedAt,
            },
            rsp: {
              caseManagerFirstName: ownerFirstName,
              name: chosenRsp.name,
            },
          },
        ),
        Mail.send(
          { to: { address: clientUser!.email, name: clientUser!.name } },
          'RspApplication',
          {
            user: newClient,
            reentryServiceProvider: chosenRsp.name,
          },
        ),
      ]);
    }
  }
}
