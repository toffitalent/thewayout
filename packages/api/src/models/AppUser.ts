import type { RelationMappingsThunk } from 'objection';
import { App } from './App';
import { BaseModel } from './BaseModel';
import { AppUserSerializer } from './serializers';
import { User } from './User';

export class AppUser extends BaseModel {
  static tableName = 'app_users';
  static Serializer = AppUserSerializer;

  readonly id!: string;
  appId!: string;
  userId!: string;
  scope!: string[];

  app?: App;
  user?: User;

  static relationMappings: RelationMappingsThunk = () => ({
    app: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: App,
      join: {
        from: 'app_users.appId',
        to: 'apps.id',
      },
    },

    user: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: User,
      join: {
        from: 'app_users.userId',
        to: 'users.id',
      },
    },
  });
}
