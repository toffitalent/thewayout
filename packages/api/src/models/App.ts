import type { RelationMappingsThunk } from 'objection';
import { AppClient } from './AppClient';
import { AppToken } from './AppToken';
import { AppUser } from './AppUser';
import { BaseModel } from './BaseModel';
import { AppSerializer } from './serializers';

export class App extends BaseModel {
  static tableName = 'apps';
  static Serializer = AppSerializer;

  readonly id!: string;
  name!: string;
  userId!: string | null;

  clients?: AppClient[];
  tokens?: AppToken[];
  users?: AppUser[];

  static relationMappings: RelationMappingsThunk = () => ({
    clients: {
      relation: BaseModel.HasManyRelation,
      modelClass: AppClient,
      join: {
        from: 'apps.id',
        to: 'app_clients.appId',
      },
    },

    tokens: {
      relation: BaseModel.HasManyRelation,
      modelClass: AppToken,
      join: {
        from: 'apps.id',
        to: 'app_tokens.appId',
      },
    },

    users: {
      relation: BaseModel.HasManyRelation,
      modelClass: AppUser,
      join: {
        from: 'apps.id',
        to: 'app_users.appId',
      },
    },
  });
}
