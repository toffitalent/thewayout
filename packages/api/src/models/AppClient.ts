import type { RelationMappingsThunk } from 'objection';
import { AuthGrantType } from '@two/shared';
import { App } from './App';
import { AppToken } from './AppToken';
import { BaseModel } from './BaseModel';

export class AppClient extends BaseModel {
  static tableName = 'app_clients';

  readonly id!: string;
  appId!: string;
  name!: string;
  secret!: string;
  grantType!: AuthGrantType;
  redirectUri!: string | null;

  app?: App;
  tokens?: AppToken[];

  static relationMappings: RelationMappingsThunk = () => ({
    app: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: App,
      join: {
        from: 'app_clients.appId',
        to: 'apps.id',
      },
    },

    tokens: {
      relation: BaseModel.HasManyRelation,
      modelClass: AppToken,
      join: {
        from: 'app_clients.id',
        to: 'app_tokens.clientId',
      },
    },
  });
}
