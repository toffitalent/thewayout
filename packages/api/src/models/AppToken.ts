import type { RelationMappingsThunk } from 'objection';
import { App } from './App';
import { AppClient } from './AppClient';
import { BaseModel } from './BaseModel';
import { User } from './User';

export class AppToken extends BaseModel {
  static tableName = 'app_tokens';

  readonly id!: string;
  appId!: string;
  clientId!: string;
  userId!: string;
  token!: Buffer;
  self!: boolean;

  app?: App;
  client?: AppClient;
  user?: User;

  static relationMappings: RelationMappingsThunk = () => ({
    app: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: App,
      join: {
        from: 'app_tokens.appId',
        to: 'apps.id',
      },
    },

    client: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: AppClient,
      join: {
        from: 'app_tokens.clientId',
        to: 'app_clients.id',
      },
    },

    user: {
      relation: BaseModel.BelongsToOneRelation,
      modelClass: User,
      join: {
        from: 'app_tokens.userId',
        to: 'users.id',
      },
    },
  });
}
