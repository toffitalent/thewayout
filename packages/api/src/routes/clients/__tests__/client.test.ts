import { faker } from '@faker-js/faker';
import { RspClient } from '@app/models/RspClient';
import { auth, fixtures, request, resetDb } from '@test';

describe('Routes > Clients > client', () => {
  beforeEach(async () => {
    await resetDb();

    faker.seed(1);
  });

  describe('DELETE /api/v1/clients/:id/rsp', () => {
    test('remove rsp client relation', async () => {
      const existingRspClient = await RspClient.query()
        .where({ userId: fixtures.user4.id })
        .first();
      expect(existingRspClient).not.toBeUndefined();

      const res = await request()
        .delete(`/api/v1/clients/${fixtures.clientProfile1.id}/rsp`)
        .use(auth(fixtures.user1.id, ['admin']));

      expect(res.status).toBe(204);
      expect(res.body).toMatchSnapshot();

      const removedRspClient = await RspClient.query().where({ userId: fixtures.user4.id }).first();
      expect(removedRspClient).toBeUndefined();
    });

    test('does not allow request from non-admins', async () => {
      const res = await request()
        .delete(`/api/v1/clients/${fixtures.clientProfile1.id}/rsp`)
        .use(auth());

      expect(res.status).toBe(403);
      expect(res.type).toBe('application/json');
      expect(res.body).toMatchSnapshot();
    });
  });

  describe('PATCH /clients/:id', () => {
    test('updates client profile', async () => {
      const res1 = await request()
        .patch(`/api/v1/clients/${fixtures.clientProfile1.id}`)
        .send({
          relativeExperience: [
            {
              startAtMonth: 'May',
              startAtYear: '2015',
              endAtMonth: '5',
              endAtYear: '2016',
            },
          ],
        })
        .use(auth(fixtures.user4.id, ['client']));
      expect(res1.status).toBe(400);

      const res2 = await request()
        .patch(`/api/v1/clients/${fixtures.clientProfile1.id}`)
        .send({
          relativeExperience: [
            {
              startAtMonth: '5',
              startAtYear: '2015',
              endAtMonth: '5',
              endAtYear: '2016',
            },
          ],
        })
        .use(auth(fixtures.user4.id, ['client']));
      expect(res2.status).toBe(200);
    });
  });
});
