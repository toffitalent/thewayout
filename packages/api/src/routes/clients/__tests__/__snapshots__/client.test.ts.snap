// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes > Clients > client DELETE /api/v1/clients/:id/rsp does not allow request from non-admins 1`] = `
{
  "code": "InsufficientScope",
  "message": "Insufficient scope has been authorized for this request",
  "status": 403,
}
`;

exports[`Routes > Clients > client DELETE /api/v1/clients/:id/rsp remove rsp client relation 1`] = `{}`;
